### ISSUE TITLE
Chat improvements

### ISSUE THREAD

**Issue by pitaj** [NodeBB/NodeBB#7330]
This is a centralized issue covering the collection of planned chat improvements

- [x] Ability to add a group. This should not be accomplished by adding every member of the group, as then new group members would not be included
- [x] Multiple chat room managers
- [x] Ability for each user to configure notifications on a per-room basis
  + Notify all members (default)
  + Notify only members currently online
  + Disable notifications
- [x] Ability to configure the default notification setting of a chat room
- [x] Chat room widget which embeds the chat room in a page
- [ ] Sent, received, read indicator (like WhatsApp etc) [#7192]
- [x] System messages for joining/leaving/renaming of a chat room

These changes will deprecate the global-chat plugin and likely replace the shoutbox widget as well. NodeBB/nodebb-plugin-global-chat#5

**Comment by youhosi** [comment]
It would also be nice to have more options in the same place to enter, e.g.

![zrzut ekranu z 2019-02-02 11-58-33](https://user-images.githubusercontent.com/24297084/52163417-ee5bc700-26e1-11e9-834d-65ac4b3c5470.png)

**Comment by pitaj** [comment]
Are you talking about formatting buttons?

**Comment by pitaj** [comment]
@PostMidnight most of those features are, in my opinion, not super useful for a forum. Better served with plugins IMO.

**Comment by PostMidnight** [comment]
Hi @pitaj perhaps just consider (1)? - its a minor improvement that could be enabled by integration of the following plugin: https://github.com/julianlam/nodebb-plugin-audio-embed.

**Comment by PatheticMustan** [comment]
One step closer to discord, am i rite ;)

**Comment by arranka** [comment]
many users say that the view in the chat cell phone has a problem
the chat notification icon is interposed with the send message button
you want to press the send button but the chat notification button is pressed
This is so? or does it have to do with the fear that he is using?

Thank you

**Comment by pitaj** [comment]
@arranka that's already been fixed in the latest persona theme. I suggest upgrading NodeBB to the latest version.

**Comment by arranka** [comment]
@pitaj updated and corrected, thanks

**Comment by PostMidnight** [comment]
IMHO the notification system can use a minor revision so it may become more real time.
May I suggest to generate action:notification.pushed event if:
(The client's current browser page is not NodeBB) AND (the user has not seen the notification within say 20 seconds from the time it was dispatched).

**Comment by PostMidnight** [comment]
Chat bubbles with different colors (for example, use avatar color).
Float bubble left/right based on sender / receiver.

See below example:
![image](https://user-images.githubusercontent.com/43993784/52650992-0c040a00-2ef4-11e9-80b4-566b91d34629.png)

**Comment by arranka** [comment]
I closed my old forum to install a nodebb forum with a virgin database, the users have tried the chat and this is what they said

In the global chat the messages that are written in the chat are not seen because it is covered by the users registered to the forum

https://i.imgur.com/GedgtWl.png

This is the view of the cell phone

https://i.imgur.com/cMc0iwn.jpg

A good idea would be that only online users will be shown in the global chat and not the users registered in the forum. Keep in mind that in the forum that closed had more than 10 thousand registered users. Imagine a list of 10 thousand registered users in the global chat and of those 10 thousand only a few are online
Maybe another good measure is to put users online on the side for example as in discord

https://i.imgur.com/qFWxun4.png

In another point users want to have the emoji button, as in discord, they also want to quote messages

When someone scrolls to read previous messages, when someone sends a new message it goes back down. Then it is impossible to read previous messages by scrolling because if someone sends a message it takes you back to the beginning. If there are 5 to 10 messages per minute, it is impossible to scroll

This is in a week of testing
Thank you very much for reading me

**Comment by pitaj** [comment]
@arranka please report that issue in the global-chat repository.

On the "only online users" idea, this has been suggested already. It's discussed in the global-chat issue linked above. It's covered in this issue by the "Notify only members currently online" item.

The solution to there being thousands of members in the global chat is not to reduce the number of users in it with naive trickery, but to fix the chat so it works correctly with many users.

**Comment by PostMidnight** [comment]
Client to client chat encryption could be useful.

**Comment by FaizanZahid** [comment]
This is the most productive issue i ever found. Let's make the node bb chat like messenger and whatsapp! This will hugely increase user retention rate on our forums. Hats off @pitaj  (Y)

**Comment by PatheticMustan** [comment]
I agree that "read/seen receipts" would be awesome, and so would chat moderators, but wouldn't audio calls and voice notes take up a lot of bandwidth/space?

**Comment by julianlam** [comment]
> but wouldn't audio calls and voice notes take up a lot of bandwidth/space?

Correct, which is why this would be plugin territory.

**Comment by PostMidnight** [comment]
Chat load time and context switching (between chats) are quite high. Perhaps it makes sense to listen to and cache all chat conversations that the user is linked to in the client, even when the user does not have a chat window open.

**Comment by PostMidnight** [comment]
@PatheticMustan read/seen can work with multiple people as well. So for instance you may flag a conversation bubble, and push an "info" icon. You may then receive a list of members who received the message, and the ones who read it, with timestamp for each.

**Comment by julianlam** [comment]
Chat loading should not actually take a long time. There's an opportunity for optimization here.

**Comment by PatheticMustan** [comment]
@PostMidnight, how long do chats normally take to load? (Just curious, I can't actually help improve code)

**Comment by PostMidnight** [comment]
@PatheticMustan, few seconds to longer timeframe if the site/network is loaded. Caching the data at the client using a silent listener should make the switch instantaneous.

**Comment by PostMidnight** [comment]
Two nice to have features:
1. Quick scroll to bottom
2. Quick search in chat content

**Comment by PatheticMustan** [comment]
NodeBB is slowly turning into Discord on a forum...

**Comment by FaizanZahid** [comment]
Discord has lots of cool features too, i hope node bb competes and be better in terms of modern features!

**Comment by arranka** [comment]
very good as it is leaving the new nodebb chat good work guys

**Comment by PatheticMustan** [comment]
Honestly all I want is the ability to add groups to chat rooms. ;(

**Comment by LonMcGregor** [comment]
I'm not sure if here is the best place to report it, but I've found a minor style bug in chat where if you use the `> quote` format for a long quote (so that it wraps multiple lines), then mobile view gets a broken width as the whole page follows the width set by the quote.

![vivaldi_2019-05-10_20-01-18](https://user-images.githubusercontent.com/3817332/57552034-7cd78880-7362-11e9-84b1-43701ad1bf35.png)

**Comment by PostMidnight** [comment]
Another nice to have feature:
View the media (pictures, video) that were shared in a group.

**Comment by PatheticMustan** [comment]
Add groups to chat would still be super cool.

**Comment by arranka** [comment]
Is the project abandoned or is it still in force?

**Comment by julianlam** [comment]
No, I don't think it's abandoned :smile:

**Comment by arranka** [comment]
with the update to 1.13.1 this error appears, guest renames chat infinitely
the chat is obsolete, cannot be used
any ideas?

![image](https://user-images.githubusercontent.com/30248115/70337087-2adbe000-1829-11ea-813a-293b927693df.png)

**Comment by DroidBV8** [comment]
now on 2021...
project abandoned?

**Comment by julianlam** [comment]
Ah, no. The project (NodeBB) is still going strong. Refactoring the chat system is just not a priority at the moment 🙂

**Comment by DroidBV8** [comment]
Hello 

Will we hope to see one day the progress points of the first post materialize.

Having quite a few users, global chat tends to be slow to open and display messages.

Currently with the number of users, my message takes 10s to appear!

**Comment by julianlam** [comment]
Indeed! It will be one of the things we improve on with the Harmony theme.

**Comment by DroidBV8** [comment]
@julina

harmony ? Not on persona ?

**Comment by julianlam** [comment]
Yes and yes, although depending on the fixes they may be backported to Persona

**Comment by julianlam** [comment]
I think most of these have been implemented. If you see any others missing, please open a separate issue for them 😄

**Comment by mySebbe** [comment]
The current checklist is effectively complete apart from entered/processed/read receipts, which already have the focused tracker #7192. The latest maintainer guidance also asks for remaining chat ideas to be reported separately. Keeping this omnibus issue open now duplicates the only known remainder, so I suggest closing it and continuing any concrete chat work in the focused issue.
